### DISCUSSION TITLE
Can I run one-time code at vite server start without a custom server?

### DISCUSSION THREAD

**Discussion by maranomynet** [remix-run/remix#8931] [Q&A] [answered]
Is it possible to run some environment validation code as the built-in Remix server starts?

I'm using the vite plugin and I'd like to avoid the maintenence burden of `remix reveal`ing a `entry.server.tsx`, and I don't otherwise need a custom server.

Related, but uses custom server: https://github.com/remix-run/remix/discussions/8694

**Comment by kiliman** [comment]
Although doable, it's kind of tricky due to the fact the development and production builds are different. In DEV mode, Vite is responsible for the web server. Whereas in PROD, we are using Remix App Server, and it uses the build output from Vite remix vite:build.

So, to tackle Vite DEV, I created a serviceStartup plugin that will let you pass in a set of services to start.

I created a folder app/startup with the startup functions.
```ts
// vite.config.ts
import serviceStart from "./app/startup/service";
import anotherServiceStart from "./app/startup/another-service";

installGlobals();

export default defineConfig({
  plugins: [
    serviceStartup({
      "my-service": serviceStart,
      "another-service": anotherServiceStart,
    }),
    remix(),
    tsconfigPaths(),
  ],
});

type ServicesType = Record<string, () => void>;

function serviceStartup(services: ServicesType) {
  const serviceStatus = new Map<string, boolean>();
  return {
    name: "service-startup",
    configureServer(server: { config: { appType: string; mode: string } }) {
      // remix vite does two builds, so only start services for custom SSR
      if (
        server.config.mode === "development" &&
        server.config.appType === "custom"
      ) {
        Object.entries(services).forEach(([name, startFunction]) => {
          // check if service already started
          if (serviceStatus.get(name)) return;
          console.log(`starting service: ${name}`);
          startFunction();
          console.log();
          serviceStatus.set(name, true);
        });
      }
    },
  };
}
```
```ts
// app/startup/service.ts
export default function start() {
  console.log("service started!");
}

// app/startup/another-service.ts
export default function start() {
  console.log("another service started!");
}
```
This should start all the services when the Vite DEV server initializes.

Now, for a PROD build, we can't update RAS, and you don't want `entry.server.tsx`. So, the next best place is the `root.tsx` route.

Add this to your root route, and it will start the services on startup.
```ts
// app/root.tsx
import serviceStart from "~/startup/service";
import anotherServiceStart from "~/startup/another-service";

function start() {
  console.log("starting services from root");
  serviceStart();
  anotherServiceStart();
}
// only run this in production, since Vite handles startup in development
if (process.env.NODE_ENV === "production") {
  start();
}
```

Normally we could just add this to root and it'll work in both DEV and PROD, but Vite doesn't execute the root route until the first request.

**Reply by kiliman** [reply]
Ugh... sorry for the edits.

**Reply by maranomynet** [reply]
Thank you so much for your detailed answer.

I've tried to both side-effect import the validation module into `root.tsx`, like so...

```ts
import './env.server'; // feeds process.env into a zod schema that throws on fail
```

…and import it as a `checkEnv` function and run it in `root.tsx`

```ts
import { checkEnv } from './env.server';

// ...
if (process.env.NODE_ENV === "production") {
  checkEnv();
}
```

but then both `yarn run remix vite:dev` and `NODE_ENV=production  yarn run remix vite:build` fail[^1] with this message:

```
Error: [commonjs--resolver] Server-only module referenced by client

    './env.server' imported by route 'app/root.tsx'

  Remix automatically removes server-code from these exports:
    `loader`, `action`, `headers`

  But other route exports in 'app/root.tsx' depend on './env.server'.

  See https://remix.run/docs/en/main/future/vite#splitting-up-client-and-server-code

    at Object.resolveId (/Users/maro5239/path/to/project/node_modules/@remix-run/dev/dist/vite/plugin.js:866:15)
```

BTW, I'm running `node@20.9.0` and `remix@2.7.2` with `vite@5.1.4`.

And I'm perfectly fine with the validation only happening on first request in development mode. It's the production server that I'm eager to see immediately fail on start.

[^1]: `vite:dev` fails when I make the first HTTP request, but `vite:build` errors mid-build.

**Reply by maranomynet** [reply]
`<peeve>`
There are so many Remix guidelines that casually involve maintaining custom `entry.server.tsx` and `entry.client.tsx`, or a custom `server.ts` just to add a single line of extra logic.

The `entry.*.tsx` files, especially, are large and full of code that _seems_ to be evolving/changing with each release of Remix.

I'd LOVE to see Remix (or the vite plugin) provide more of a middle-ground between "pure default behavior" and "go maintain your own request handler".
`</peeve>`

**Reply by maranomynet** [reply]
> `   But other route exports in 'app/root.tsx' depend on './env.server'.`

Just to be clear: There's no exported code in `root.tsx` that references anything inside  `./env.server` in any way whatsover.

And this is `env.server.ts`:
```ts
import { TypeOf, z } from 'zod';

// Ensure at runtime that all required environment variables are defined.
const url = /*#__PURE__*/ z.string().url();
const str = /*#__PURE__*/ z.string().min(1);

const envFileSchema = /*#__PURE__*/ z.object({
  API_HOST: url,
  API_KEY: str,
});

declare global {
  namespace NodeJS {
    interface ProcessEnv extends Readonly<TypeOf<typeof envFileSchema>> {}
  }
}

const checkEnv = () => {
  try {
    envFileSchema.parse(process.env);
  } catch (err) {
    if (err instanceof z.ZodError) {
      const { fieldErrors } = err.flatten();
      const errorMessage = Object.entries(fieldErrors)
        .map(([field, errors]) => (errors ? `${field}: ${errors.join(', ')}` : field))
        .join('\n  ');
      const pluralSuffix = Object.keys(fieldErrors).length > 1 ? 's' : '';
      throw new Error(`Missing environment variable${pluralSuffix}:\n  ${errorMessage}`);
      process.exit(1);
    }
  }
};

/*#__PURE__*/ checkEnv();
```

(I added the "PURE" markers in case it might help the treeshaker's confidence.)

**Reply by maranomynet** [reply]
I'm tempted to add a TS-to-JS build step to transpile env.server.ts as a standalone script and then run it before envoking `remix-serve ./build/server/index.js`.

It's just more faffin than I was hoping for.

**Reply by kiliman** [reply, marked as answer]
Ok, I figured out a solution. Install the Vite plugin `vite-env-only` by @pcattori.
```bash
npm install vite-env-only
```
Update your _vite.config.ts_ file to add the plugin.
```ts
// vite.config.ts
import envOnly from "vite-env-only";

export default defineConfig({
  plugins: [envOnly(), remix(), tsconfigPaths()],
});
```
Export your `checkEnv` function
```ts
// env.server.ts
export const checkEnv = () => {
  console.log("checking env");
};
```
Then in your root, import the function and wrap in `serverOnly$` call
```ts
// root.tsx
import { serverOnly$ } from "vite-env-only";
import { checkEnv } from "./env.server";
serverOnly$(checkEnv());
```

This will call `checkEnv()` at startup for production builds, and on first route access for development.

**Reply by maranomynet** [reply]
Ah! `serverOnly$` ...of course!

Curses for not figuring this out on my own.

Thanks!

**Reply by davidsonsns** [reply]
hey @kiliman, thanks so much for providing such a detailed explanation. Was looking for this for a while, you helped me a lot.

question: I have exactly the same structure, but for some reason it's not calling `checkEnv` on startup in production. not sure if this is a remix version thing (installed version `2.8.1`). any guesses?

btw, `vite-env-only` version `3.0.1` changed `envOnly` to `envOnlyMacros`. this could be the reason why I'm having another experience (not sure if the core of this function was revamped)

**Comment by bocajji** [comment]
A little late to the party, but what about verifying the environment variables in a separate script before starting the app? That way you avoid the issues of when verification is actually called when starting the remix server.

for example in package.json: 
```
    "verify-env-vars": "node scripts/your-env-verification-script.js",
     "dev": "npm run verify-env-vars && remix vite:dev"
```

**Reply by gustavopch** [reply]
Here, I do it like `env('MY_VAR')` and this function throws if it's required but missing.

**Reply by maranomynet** [reply]
Yes this is more feasible now that Remix/react-router works with bun, and node.js can be coerced into running .ts files directly.

In fact I think I will update my projects to do that.

Before, this would have required a transpilation step, or maintaining a separate JavaScript env validation module (providing no type safety).
